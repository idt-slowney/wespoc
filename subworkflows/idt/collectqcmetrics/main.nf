include { PICARD_COLLECTMULTIPLEMETRICS } from '../../../modules/idt/picard/collectmultiplemetrics/main'
include { PICARD_COLLECTHSMETRICS       } from '../../../modules/idt/picard/collecthsmetrics/main'
include { SENTIEON_DATAMETRICS          } from '../../../modules/idt/sentieon/datametrics/main'
// include { SENTIEON_HSMETRICS            } from '../../../modules/idt/sentieon/hsmetrics/main'
include { PICARD_COLLECTOXOGMETRICS     } from '../../../modules/idt/picard/collectoxogmetrics/main'
include { PICARD_COLLECTSEQUENCINGARTIFACTMETRICS } from '../../../modules/idt/picard/collectsequencingartifactmetrics/main'

workflow COLLECT_QC_METRICS {
    take:
    bam_and_bai  // channel: [ val(meta), [bam], [bai] ]
    bait_interval
    target_interval
    fasta                // channel: [ val(meta), fasta ]
    fasta_fai            // channel: [ val(meta), fasta_fai ]
    fasta_dict           // channel: [ val(meta), fasta_dict ]

    main:
    versions = Channel.empty()
    coverage_metrics = Channel.empty()
    multiple_metrics = Channel.empty()
    oxog_metrics = Channel.empty()
    artifact_metrics = Channel.empty()

    // -------------------------------- MultipleMetrics -------------------------------- //
    // GCBias, MeanQualityByCycle, QualityDistribution, InsertSizeMetrics, AlignmentStats
    PICARD_COLLECTMULTIPLEMETRICS( bam_and_bai, fasta, fasta_fai ) // ext.when = { accelerate_metrics_flag == false }
    // SENTIEON_DATAMETRICS( bam_and_bai, fasta, fasta_fai ) // ext.when = { accelerate_metrics_flag == true }

    multiple_metrics = multiple_metrics.mix(PICARD_COLLECTMULTIPLEMETRICS.out.metrics)
    // multiple_metrics = multiple_metrics.mix(SENTIEON_DATAMETRICS.out.metrics)
    versions = versions.mix(PICARD_COLLECTMULTIPLEMETRICS.out.versions.first())
    // versions = versions.mix(SENTIEON_DATAMETRICS.out.versions.first())

    // ----------------------------------- HsMetrics ----------------------------------- //
    // ch_hsMetrics = bam_and_bai.map{ meta, bam, bai -> [ meta, bam, bai, bait_interval, target_interval ] }
    ch_hsMetrics = bam_and_bai.combine(bait_interval).combine(target_interval)
    ch_hsMetrics.view()

    PICARD_COLLECTHSMETRICS( ch_hsMetrics, fasta, fasta_fai, fasta_dict )
    // SENTIEON_HSMETRICS()

    versions = versions.mix(PICARD_COLLECTHSMETRICS.out.versions.first())
    // versions = versions.mix(SENTIEON_HSMETRICS.out.versions.first())
    coverage_metrics = coverage_metrics.mix(PICARD_COLLECTHSMETRICS.out.metrics)
    // coverage_metrics = coverage_metrics.mix(SENTIEON_HSMETRICS.out.metrics)

    // ----------------------------------- SequenceArtifacts ----------------------------------- //
    PICARD_COLLECTOXOGMETRICS( bam_and_bai.map{ meta, bam, bai -> [meta, bam] }, fasta)
    oxog_metrics = oxog_metrics.mix(PICARD_COLLECTOXOGMETRICS.out.metrics)
    versions = versions.mix(PICARD_COLLECTOXOGMETRICS.out.versions)

    // ----------------------------------- OxoMetrics ----------------------------------- //
    PICARD_COLLECTSEQUENCINGARTIFACTMETRICS( bam_and_bai.map{ meta, bam, bai -> [meta, bam] }, fasta )
    artifact_metrics = artifact_metrics.mix( PICARD_COLLECTSEQUENCINGARTIFACTMETRICS.out.metrics )
    versions = versions.mix( PICARD_COLLECTSEQUENCINGARTIFACTMETRICS.out.versions )

    emit:
    coverage_metrics    = coverage_metrics                       // channel: [ val(meta), [ coverage_metrics ] ]
    multiple_metrics    = PICARD_COLLECTMULTIPLEMETRICS.out.metrics // channel: [ val(meta), [ multiple_metrics ] ]
    oxog_metrics        = oxog_metrics
    artifact_metrics    = artifact_metrics
    versions            = versions                               // channel: [ versions.yml ]
}
