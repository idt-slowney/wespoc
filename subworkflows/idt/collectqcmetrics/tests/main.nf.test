// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
// nf-core subworkflows test collectqcmetrics
nextflow_workflow {

    name "Test Subworkflow COLLECT_QC_METRICS"
    script "../main.nf"
    workflow "COLLECT_QC_METRICS"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/collectqcmetrics"
    tag "picard/collectmultiplemetrics"
    tag "picard/collecthsmetrics"
    tag "sentieon/datametrics"


    test("Base Test") {

        when {
            workflow {
                """
                // TODO nf-core: define inputs of the workflow here. Example:
                input[0] = Channel.value([
                    [ id:'test', single_end:false ], // meta map
                    file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExists: true),
                    file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExists: true)
                ])
                input[1] = file(params.test_data['homo_sapiens']['genome']['genome_interval_list'], checkIfExists: true)
                input[2] = file(params.test_data['homo_sapiens']['genome']['genome_interval_list'], checkIfExists: true)
                input[3] = Channel.of([ 
                    [ id:'genome' ],
                      file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true)
                ])
                input[4] = Channel.of([ 
                    [ id:'genome_fai' ],
                      file(params.test_data['homo_sapiens']['genome']['genome_fasta_fai'], checkIfExists: true)
                ])
                input[5] = Channel.of([ 
                    [ id:'genome_dict' ],
                      file(params.test_data['homo_sapiens']['genome']['genome_dict'], checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
                //TODO nf-core: Add all required assertions to verify the test output.
            )
        }
    }
}
