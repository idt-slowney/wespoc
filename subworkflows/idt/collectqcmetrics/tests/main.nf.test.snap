{
    "Base Test": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.CollectHsMetrics.coverage_metrics:md5,d97e73ebd635c8831c4eccae67169d7b"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test.CollectMultipleMetrics.alignment_summary_metrics:md5,1aa8b0f3bd84282631786e6daddeb728",
                            "test.CollectMultipleMetrics.base_distribution_by_cycle_metrics:md5,1ede0c669af457d64e0e85ad31532887",
                            "test.CollectMultipleMetrics.insert_size_metrics:md5,11c89714b33cc5c6841393eb1350ce59",
                            "test.CollectMultipleMetrics.quality_by_cycle_metrics:md5,9132a97992f054d1c319d7f3e09107f3",
                            "test.CollectMultipleMetrics.quality_distribution_metrics:md5,a33744018b03c47f56322be31a1ec5b1"
                        ]
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_OxoGMetrics.txt:md5,ad053db575f8c436d860d6ea87255ded"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.error_summary_metrics.error_summary_metrics:md5,737ec16b56ca7e5b11d89eb3d724af30"
                    ]
                ],
                "4": [
                    "versions.yml:md5,8c81422b114d007b53b2a987ed6f220f",
                    "versions.yml:md5,bcb3347e16a63f8c141261c5adcb40d3",
                    "versions.yml:md5,ca596aebc9796bd5119536afceabf58d",
                    "versions.yml:md5,e6eff9debb86fa441f945c1d07936f78"
                ],
                "artifact_metrics": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.error_summary_metrics.error_summary_metrics:md5,737ec16b56ca7e5b11d89eb3d724af30"
                    ]
                ],
                "coverage_metrics": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.CollectHsMetrics.coverage_metrics:md5,d97e73ebd635c8831c4eccae67169d7b"
                    ]
                ],
                "multiple_metrics": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test.CollectMultipleMetrics.alignment_summary_metrics:md5,1aa8b0f3bd84282631786e6daddeb728",
                            "test.CollectMultipleMetrics.base_distribution_by_cycle_metrics:md5,1ede0c669af457d64e0e85ad31532887",
                            "test.CollectMultipleMetrics.insert_size_metrics:md5,11c89714b33cc5c6841393eb1350ce59",
                            "test.CollectMultipleMetrics.quality_by_cycle_metrics:md5,9132a97992f054d1c319d7f3e09107f3",
                            "test.CollectMultipleMetrics.quality_distribution_metrics:md5,a33744018b03c47f56322be31a1ec5b1"
                        ]
                    ]
                ],
                "oxog_metrics": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_OxoGMetrics.txt:md5,ad053db575f8c436d860d6ea87255ded"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8c81422b114d007b53b2a987ed6f220f",
                    "versions.yml:md5,bcb3347e16a63f8c141261c5adcb40d3",
                    "versions.yml:md5,ca596aebc9796bd5119536afceabf58d",
                    "versions.yml:md5,e6eff9debb86fa441f945c1d07936f78"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-27T15:52:07.340844444"
    }
}